app/src/main/java/org/sagebionetworks/research/mpower/tracking/view_model/TrackingTaskViewModel.java

```diff
@@ -29,63 +36,154 @@
  */
 public abstract class TrackingTaskViewModel<ConfigType extends TrackingItemConfig, LogType extends TrackingItemLog>
         extends ViewModel {
-    protected LiveData<Boolean> selectionMade;
+    private static final Logger LOGGER = LoggerFactory.getLogger(TrackingTaskViewModel.class);
+
+    public static final String LOGGING_COLLECTION_IDENTIFIER = "trackedItems";
 
     // invariant availableElements.getValue() != null
+    @NonNull
     protected MutableLiveData<Map<TrackingSection, Set<TrackingItem>>> availableElements;
 
-    // invariant activeElements.getValue() != null
-    protected MutableLiveData<Set<ConfigType>> activeElements;
+    @NonNull
+    protected Map<String, TrackingItem> trackingItemsById;
+
+    @NonNull
+    // invariant activeElementsById.getValue() != null
+    protected MutableLiveData<Map<String, ConfigType>> activeElementsById;
+
+    @NonNull
+    // invariant loggedElementsById.getValue() != null
+    protected MutableLiveData<Map<String, LogType>> loggedElementsById;
 
-    // invariant unconfiguredElements.getValue() != null
-    protected LiveData<Set<ConfigType>> unconfiguredElements;
+    protected Instant startDate;
 
-    // invariant loggedElements.getValue() != null
-    protected MutableLiveData<Set<LogType>> loggedElements;
+    protected Instant endDate;
 
+    @NonNull
     protected TrackingStepView stepView;
 
-    protected TrackingTaskViewModel(@NonNull final TrackingStepView stepView) {
+    protected TrackingTaskViewModel(@NonNull final TrackingStepView stepView,
+            @Nullable final LoggingCollection<LogType> previousLoggingCollection) {
         this.stepView = stepView;
-        this.availableElements = new MutableLiveData<>();
-        this.availableElements.setValue(stepView.getSelectionItems());
-        this.activeElements = new MutableLiveData<>();
-        this.activeElements.setValue(new HashSet<>());
-        this.selectionMade = Transformations.map(this.activeElements, (elements) -> !elements.isEmpty());
-        this.unconfiguredElements = Transformations.map(this.activeElements, (elements) -> {
-            Set<ConfigType> result = new HashSet<>();
-            for (ConfigType config : elements) {
-                if (!config.isConfigured()) {
-                    result.add(config);
-                }
+        availableElements = new MutableLiveData<>();
+        activeElementsById = new MutableLiveData<>();
+        loggedElementsById = new MutableLiveData<>();
+        availableElements.setValue(stepView.getSelectionItems());
+        trackingItemsById = getTrackingItemsById(availableElements.getValue());
+        loggedElementsById.setValue(new HashMap<>());
+        // initialize the active elements to contain either the user's previous selection or nothing depending on if
+        // we have a previous logging collection.
+        if (previousLoggingCollection == null) {
```

Comment:
Isn't the previousLoggingCollection the responsibility of this ViewModel to determine?

Reply:
On iOS it is based on the most recently finished study report... Which I guess just isn't implemented yet, so this is fine for now.  Maybe put in a TODO

Reply:
I was wondering about that. The problem is that I don't want the view model to have to get the study report since it may make the fragments laggy as they wait for the view model to get the report. I'll leave a TODO for now.